Adding spoken dialogue state tracking recipes for Spoken MultiWoz and SpokenWoz datasets

LucasDruart opened the pull request:
## What does this PR do?

This PR adds spoken dialogue state tracking recipes for two datasets: spoken MultiWoz and SpokenWoz. I have tried my best to conform to speechbrain's conventions. Please let me know if there are any conventions I missed. I do not know if and how recipes should be tested.

<details>
  <summary><b>Before submitting</b></summary>

- [x] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [x] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [x] Review the self-review checklist to ensure the code is ready for review

</details>

<!--

🎩 Magic happens when you code. Keep the spells flowing!

-->

asumagic commented:
#2441 also adds a T5 HuggingFace lobe as far as I could tell, we will see how we deal with this.

LucasDruart commented:
Hey @asumagic, 
Thanks for taking a look into this PR, I have seen that the pre-commit pipeline fails because of the spokenwoz data preparation function which is too complicated. I believe the complexity is measured in number of lines of code, however I have trouble simplifying it since the formatting sometimes reformats into many lines. For instance [line 146](https://github.com/LucasDruart/speechbrain/blob/dialogue_state_tracking/recipes/SpokenWoz/dialogue_state_tracking/spokenwoz_prepare_slu.py#L146).

asumagic commented:
> Hey @asumagic, Thanks for taking a look into this PR, I have seen that the pre-commit pipeline fails because of the spokenwoz data preparation function which is too complicated. I believe the complexity is measured in number of lines of code, however I have trouble simplifying it since the formatting sometimes reformats into many lines. For instance [line 146](https://github.com/LucasDruart/speechbrain/blob/dialogue_state_tracking/recipes/SpokenWoz/dialogue_state_tracking/spokenwoz_prepare_slu.py#L146).  

I think the linter measures [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity) rather than the plain number of lines of code, so fixing formatting would probably not help. In this case, it could make sense to extract the whole `for dialog_id, dialog_info in annotations.items():` loop into its own function.

As for the other pre-commit style check fails, you can fix them locally by running `pre-commit run -a` ([pypi](https://pypi.org/project/pre-commit/)).

As for the other tests failing, the issue seems to be that recipe tests should be added. Recipe tests use a small subset of data and launches a dummy training for a fixed number of epochs, then asserting that some metric is respected.  
(They are **not** run inside of CI; but rather run by SpeechBrain developers locally.)  
As far as I can tell, a line should be added in `tests/recipes/MultiWOZ.csv` and a `tests/recipes/SpokenWOZ.csv` file should be created. Recipe tests are further explained in [`tests/recipes`](https://github.com/speechbrain/speechbrain/tree/develop/tests/recipes).

LucasDruart commented:
Hi @asumagic,
I fixed the complexity and trailing space failed tests and added a test for each recipe. I had to commit without verifications the last commit since I had a failed test for large files for the sample audio (tests/samples/DST/SpokenWoz/audio_5700_train_dev/MUL1011.wav) required for the test. Let me know if I can do anything else.

asumagic commented:
BTW, I am still in the process of reviewing, but it does take some time.

asumagic commented on speechbrain/utils/evaluate_dialogue_state_tracking.py:
```diff
@@ -0,0 +1,685 @@
+        )
+
+
+class DSTMetrics:
```
IMO should be moved to `metric_stats.py` without any dataset-specific things (including defaults) and made a child of `MetricStats`.

asumagic commented on speechbrain/utils/evaluate_dialogue_state_tracking.py:
```diff
@@ -0,0 +1,685 @@
+]
+
+
+def dialogueState_str2dict(
```
Should be renamed `dialogue_state_str2dict` for consistency.

asumagic commented on speechbrain/utils/evaluate_dialogue_state_tracking.py:
```diff
@@ -0,0 +1,685 @@
+    return "; ".join(slots)
+
+
+class JGATracker:
```
Should be moved to `metric_stats.py` and made a child class of `MetricStats` (see the WER stats class for reference).

asumagic commented on speechbrain/utils/evaluate_dialogue_state_tracking.py:
```diff
@@ -0,0 +1,685 @@
+                self.correct += 1
+            self.total += 1
+
+    def summarize(self):
```
Document return type

asumagic commented on speechbrain/utils/evaluate_dialogue_state_tracking.py:
```diff
@@ -0,0 +1,685 @@
+#!/usr/bin/env python3
+"""
```
I feel like the metrics should be extracted into `metric_stats.py` (as mentioned in a later comment) with any dataset-specific quirks as part of the recipes under `meta/`.

If you can think of a clean way to specify all the dataset-specific stuff that isn't centralized that would avoid duplicating the invocation scripts, feel free (in which case it should probably be under `tools` not `utils`).

I couldn't really test this part myself so I refrained from making heavier changes here.

asumagic commented on speechbrain/utils/evaluate_dialogue_state_tracking.py:
```diff
@@ -0,0 +1,685 @@
+        return dict_state
+
+
+def dialogueState_dict2str(dialogue_state: dict):
```
Same comment on naming

asumagic commented on speechbrain/utils/evaluate_dialogue_state_tracking.py:
```diff
@@ -0,0 +1,685 @@
+                100 * total_correct / total, 1
+            )
+
+    def get_average_jga(self):
```
Document return type

asumagic reviewed (changes requested):
Tried to review as much as I could but it remained less in-depth as I would have liked since I am not yet very familiar with this task. I have successfully run the dataset download steps and data preparation and run some steps of training so this part works OK.  
I also made a number of changes you've probably noticed, let me know if they seem sane to you. The tests now pass.

There are some things I couldn't test and run (inference) as I don't think you have provided pre-trained models. If you are able to upload them, we could host it on the Dropbox and have it linked.

I would like other reviewers to also double-check the training script + metrics for correctness.

As for the duplication across training scripts (and hparams), I agree it isn't pretty, but I think this is just a flaw we have right now where very similar models just result in duplication across recipes. IMO this can be dealt with later if we figure out how to deduplicate dataset-agnostic things better.

LucasDruart commented on speechbrain/decoders/seq2seq.py:
```diff
@@ -265,7 +265,7 @@ def _get_top_prediction(self, hyps, scores, log_probs):
 
         # Use SpeechBrain style lengths
-        top_lengths = (top_lengths - 1).abs() / max_length
+        top_lengths = (top_lengths).abs() / max_length
```
I noticed the last token was always missing in the predictions. I believe that the `top_lengths - 1` is supposed to remove the eos, however `(pred == self.eos_index).nonzero(as_tuple=False)` on line 256 already does it. I am not sure rather it alters behavior of other things.

LucasDruart replied:
Is my understanding correct @Adel-Moumen and do you believe it alters the behavior of other things?

LucasDruart commented:
Sorry for the delay and thank you for your review and changes which make a lot of sense! Regarding the other points:
- I have not provided checkpoints but can provide them if needed. However it requires to re-run the whole training to reach the reported performances.
- I corrected the few naming issues and missing documentation, hope its okay now.
- I moved the Joint-Goal Accuracy tracker (used for training to keep a running JGA without storing the values) to `metric_stats.py`.
- I left the common part of Dialogue State Tracking evaluation in `utils` to be able to import it from the recipes.
- I moved the dataset specific evaluations to their recipe's meta folder.

Let me know what else I can do to go forward with this PR.
